[Datasets] Remove "ML Preprocessing" user guide

bveeramani opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

The "ML Preprocessing" guide contains outdated or redundant information. With the introduction of the "Using Preprocessors" and "Configuring Training Datasets" guides, the page is largely obsolete. This PR removes "ML Preprocessing" and moves still-relevant information to other guides.

## Related issue number

<!-- For example: "Closes #1234" -->

Closes https://github.com/ray-project/ray/issues/30410

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

angelinalg commented on doc/source/data/dataset.rst:
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
-Ray Datasets are not intended as a replacement for more general data processing systems.
-:ref:`Learn more about how Ray Datasets work with other ETL systems <datasets-ml-preprocessing>`.
+Datasets is not intended as a replacement for more general data processing systems.
+Its utility is as the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray:
```
```suggestion

```

angelinalg replied:
If you're comfortable being more opinionated about this guidance. Think of how you would tell a friend who is using Datasets.

Also, do we need to explain what we mean by `last-mile`? I thought that might be a term that we coined internally. If not, no need to explain.

bveeramani replied:
Hmm...really not sure how I'd describe Datasets' utility in a simple but precise way.

Maybe something like "Datasets moves data from storage to distributed Ray applications. It also lets you transform data into representations that are more suitable for models."?

c21 replied:
Can we remove "Its utility is as the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray:".

This is pretty much repeated from above sentence: "Use Datasets for the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray."

We can add a link to "User Guides" here, e.g. "Check out Datasets User Duides for the features it supports".

angelinalg commented on doc/source/data/dataset.rst:
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
+..
+  https://docs.google.com/presentation/d/1l03C1-4jsujvEFZUM4JVNy8Ju8jnY5Lc_3q7MBWi2PQ/edit
+
+Ray-integrated DataFrame libraries can also be seamlessly used with Datasets, to enable running a full data to
```
```suggestion
You can also use Datasets seamlessly with Ray-integrated DataFrame libraries. You can run a data to
```

angelinalg commented on doc/source/data/dataset.rst:
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
+  https://docs.google.com/presentation/d/1l03C1-4jsujvEFZUM4JVNy8Ju8jnY5Lc_3q7MBWi2PQ/edit
+
+Ray-integrated DataFrame libraries can also be seamlessly used with Datasets, to enable running a full data to
+ML pipeline completely within Ray without requiring data to be materialized to external storage:
```
```suggestion
ML pipeline completely within Ray without materializing data to external storage:
```

angelinalg replied:
Not sure if I mangled the intended meaning, but I'm trying to avoid some passive voice.

angelinalg commented on doc/source/data/transforming-datasets.rst:
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
   When using the actor compute strategy, per-row and per-batch UDFs can also be
```
```suggestion
  With the actor compute strategy, you can use per-row and per-batch UDFs
```

angelinalg commented on doc/source/data/transforming-datasets.rst:
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
   When using the actor compute strategy, per-row and per-batch UDFs can also be
   *callable classes*, i.e. classes that implement the ``__call__`` magic method. The
```
```suggestion
  *callable classes*, i.e., classes that implement the ``__call__`` magic method. The
```

angelinalg commented on doc/source/data/transforming-datasets.rst:
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
   When using the actor compute strategy, per-row and per-batch UDFs can also be
   *callable classes*, i.e. classes that implement the ``__call__`` magic method. The
   constructor of the class can be used for stateful setup, and will be only invoked once
```
```suggestion
  can use the constructor of the class for stateful setup, and it is only invoked once
```

angelinalg commented on doc/source/data/transforming-datasets.rst:
```diff
@@ -113,8 +113,8 @@ There are three types of UDFs that you can use with Ray Data: Function UDFs, Cal
-  
+  per worker actor.
+
   Callable classes are useful if there is expensive state (such as a neural network) that need to be loaded for the UDF. By using an actor class, the state only needs to be loaded once in the beginning, rather than for each batch.
```
```suggestion
  Callable classes are useful if you need to load an expensive state (such as a neural network) for the UDF. By using an actor class, the you only need to load the state once in the beginning, rather than for each batch.
```

angelinalg commented on doc/source/data/dataset.rst:
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
 Ray Datasets are designed to load and preprocess data for distributed :ref:`ML training pipelines <train-docs>`.
```
```suggestion
Use Ray Datasets to load and preprocess data for distributed :ref:`ML training pipelines <train-docs>`.
```

angelinalg commented on doc/source/data/dataset.rst:
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
 
-Ray Datasets are not intended as a replacement for more general data processing systems.
-:ref:`Learn more about how Ray Datasets work with other ETL systems <datasets-ml-preprocessing>`.
+Datasets is not intended as a replacement for more general data processing systems.
```
```suggestion
Use Datasets for the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray. Do not use them as a replacement for more general data processing systems.
```

angelinalg commented:
Very excited about the maintenance work you're doing on the docs!

angelinalg commented on doc/source/data/dataset.rst:
```diff
@@ -36,8 +36,22 @@ Data Loading and Preprocessing for ML Training
+  https://docs.google.com/presentation/d/1l03C1-4jsujvEFZUM4JVNy8Ju8jnY5Lc_3q7MBWi2PQ/edit
+
+Ray-integrated DataFrame libraries can also be seamlessly used with Datasets, to enable running a full data to
+ML pipeline completely within Ray without requiring data to be materialized to external storage:
```
```suggestion
ML pipeline completely within Ray without requiring data to be materialized to external storage.
```

amogkam commented:
Thanks @bveeramani. 

1. the random shuffling/randomize block order section is being dropped right? Where should I be looking at for this information?
2. Can we create more synergy with the AIR docs? can we link to the AIR preprocessing pages and Check ingest pages as appropriate? For example, refer to the AIR preprocessing doc in the "Transforming Datasets" section

bveeramani commented:
> 1. the random shuffling/randomize block order section is being dropped  in this PR? Where should I be looking at for this information?

[Shuffling Data](https://docs.ray.io/en/latest/ray-air/check-ingest.html#shuffling-data).

> 2. Can we create more synergy with the AIR docs? can we link to the AIR preprocessing pages and Check ingest pages as appropriate? For example, refer to the AIR preprocessing doc in the "Transforming Datasets" section

Yeah, will add.

amogkam commented:
It's not clear I need to go to AIR docs to see how to do shuffling

clarkzinzow commented on doc/source/ray-air/check-ingest.rst:
```diff
@@ -131,7 +131,13 @@ Enabling Streaming Ingest
-Shuffling or data randomization is important for training high-quality models. By default, AIR will randomize the order the data files (blocks) are read from. AIR also offers options for further randomizing data records within each file:
+Shuffling or data randomization is important for training high-quality models.
+
+By default, AIR reads data files (blocks) in a random order and shuffles the blocks
+assigned to each dataset shard. You can disable this behaviour by setting
```
```suggestion
By default, AIR shuffles the assignment of data blocks (files) to dataset shards between epochs. You can disable this behavior by setting
```

clarkzinzow replied:
Also not sure if we should call out how to disable this, since this is a very cheap operation that is an easy win for improving training data shuffling. Is there a use case in which you'd want to disable this? If not, I don't think we need a disabling call out.

bveeramani replied:
If you're evaluating a model on a validation set, you wouldn't want to shuffle data blocks.

clarkzinzow commented on doc/source/data/dataset.rst:
```diff
@@ -33,11 +33,24 @@ is already supported.
 
-Ray Datasets are not intended as a replacement for more general data processing systems.
-:ref:`Learn more about how Ray Datasets work with other ETL systems <datasets-ml-preprocessing>`.
+Use Datasets for the last-mile bridge from ETL pipeline outputs to distributed applications and libraries in Ray. Do not use them as a replacement for more general data processing systems.
```
```suggestion
Use Datasets as a last-mile bridge from storage or ETL pipeline outputs to distributed applications and libraries in Ray. Do not use it as a replacement for more general data processing systems.
```

c21 commented on doc/source/data/dataset.rst:
```diff
@@ -33,11 +33,24 @@ is already supported.
+..
+  https://docs.google.com/presentation/d/1l03C1-4jsujvEFZUM4JVNy8Ju8jnY5Lc_3q7MBWi2PQ/edit
+
+You can also use Datasets seamlessly with Ray-integrated DataFrame libraries.
+
+.. image:: images/dataset-loading-2.png
+   :width: 650px
+   :align: center
```
I am thinking of to drop this part. If we look at the images, they are mostly identical, except the scope of "Ray cluster". And to run Spark/Mars/etc on Ray is community-maintained, and not battle tested. So it should be safe to drop here.

bveeramani replied:
Removed

clarkzinzow commented:
+1 to more cross links from Datasets docs to AIR docs.

As for the shuffling docs, why don't we add a shuffle section to the "Transforming Datasets" feature guide that only details the `ds.random_shuffle()` API (basically the section that we have in the current "ML Preprocessing" guide), and have it cross-link to the AIR docs on shuffling + call out that local shuffling is a good alternative, but delegate details to the AIR level docs?

c21 commented on doc/source/data/transforming-datasets.rst:
```diff
@@ -373,6 +373,97 @@ The following output types are allowed for per-row UDFs (e.g.,
     :start-after: __writing_simple_out_row_udfs_begin__
     :end-before: __writing_simple_out_row_udfs_end__
 
+.. _datasets-groupbys:
```
Shall we put this section to the end of this file?

`Writing User-defined Functions (UDFs)` -> `Configuring Batch Size` -> `Compute Strategy` is a coherent story for writing `map_batches`.

c21 commented on doc/source/data/transforming-datasets.rst:
```diff
@@ -373,6 +373,97 @@ The following output types are allowed for per-row UDFs (e.g.,
     :end-before: __writing_simple_out_row_udfs_end__
 
+.. _datasets-groupbys:
+
```
what about https://docs.ray.io/en/master/data/dataset-ml-preprocessing.html#random-shuffle and https://docs.ray.io/en/master/data/dataset-ml-preprocessing.html#random-block-order? Should we add here as well?

bveeramani replied:
For `randomize_block_order`, I've added a callout in "Shuffling Data" section of the AIR user guide.

For `random_shuffle`, I'll follow Clark's suggestion:
> As for the shuffling docs, why don't we add a shuffle section to the "Transforming Datasets" feature guide that only details the ds.random_shuffle() API (basically the section that we have in the current "ML Preprocessing" guide), and have it cross-link to the AIR docs on shuffling + call out that local shuffling is a good alternative, but delegate details to the AIR level docs?

bveeramani commented:
> 1. the random shuffling/randomize block order section is being dropped  in this PR? Where should I be looking at for this information?

Added a "Shuffling data" section to "Transforming data" and linked to the "Shuffling data" section of the AIR user guide.

> 2. Can we create more synergy with the AIR docs? can we link to the AIR preprocessing pages and Check ingest pages as appropriate? For example, refer to the AIR preprocessing doc in the "Transforming Datasets" section

In addition to linking to the AIR docs in "Shuffling data", I also added a reference to the top of "Transforming Datasets" that points to "Using Preprocessors".

angelinalg reviewed (approved):
LGTM

c21 reviewed (approved):
LG
